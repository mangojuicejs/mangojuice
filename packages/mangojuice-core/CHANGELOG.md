# Change Log

All notable changes to this project will be documented in this file.
See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.1.3"></a>
## 0.1.3 (2017-07-12)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))
* **core:** recursive nesting works now ([c462be5](https://github.com/mangojuicejs/mangojuice/commit/c462be5))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.3"></a>
## 0.1.3 (2017-07-08)



<a name="0.1.3"></a>
## 0.1.3 (2017-07-08)



<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.3"></a>
## 0.1.3 (2017-07-08)



<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)



<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)
