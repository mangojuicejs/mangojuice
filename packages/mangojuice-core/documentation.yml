toc:
  - name: Basic
    description: |
      The most common APIs that you will work with most of the time
    children:
      - Block
      - LogicBase
      - cmd
      - run
      - mount
      - task
      - cancel
      - child
      - depends
  - name: Extras
    description: |
      Other pieces of MangoJuice
    children:
      - Process
      - Command
      - TaskMeta
      - DefaultLogger
      - ComputedField
      - Mounter
      - defineCommand
      - ensureCommand
      - decorateLogic
      - handleBefore
      - handleAfter
      - handle
      - call
      - logicOf
      - observe
      - procOf
      - bind
      - hydrate
      - delay
