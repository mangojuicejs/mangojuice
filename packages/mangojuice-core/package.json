{
  "name": "mangojuice-core",
  "version": "0.1.6",
  "description": "",
  "author": {
    "name": "Artem Artemev",
    "email": "viiv.c58@gmail.com"
  },
  "bugs": {
    "url": "https://github.com/mangojuicejs/mangojuice/issues"
  },
  "homepage": "https://github.com/mangojuicejs/mangojuice#readme",
  "repository": "https://github.com/mangojuicejs/mangojuice",
  "license": "MIT",
  "main": "index.js",
  "rollup": {
    "moduleName": "Mjs"
  }
}
