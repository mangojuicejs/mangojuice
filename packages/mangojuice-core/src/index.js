import Process from './classes/Process';
import Command from './classes/Command';
import LogicBase from './classes/LogicBase';
import DefaultLogger from './classes/DefaultLogger';
import cmd from './core/cmd/cmd';
import cancel from './core/cmd/cancel';
import defineCommand from './core/cmd/defineCommand';
import ensureCommand from './core/cmd/ensureCommand';
import child from './core/logic/child';
import depends from './core/logic/depends';
import observe from './core/logic/observe';
import { handle, handleAfter, handleBefore } from './core/logic/handle';
import logicOf from './core/logic/logicOf';
import procOf from './core/logic/procOf';
import decorateLogic from './core/logic/decorateLogic';
import delay from './core/task/delay';
import task from './core/task/task';
import callTask from './core/task/callTask';
import { CANCEL } from './classes/AsyncTask';
import run from './core/run/run';
import bind from './core/run/bind';
import mount from './core/run/mount';
import hydrate from './core/run/hydrate';
import * as utils from './core/utils';
import * as config from './config';


export {
  // Constants
  CANCEL,
  // Classes
  Process,
  Command,
  LogicBase,
  DefaultLogger,
  // Core
  cmd,
  cancel,
  defineCommand,
  ensureCommand,
  child,
  depends,
  observe,
  handle,
  handleAfter,
  handleBefore,
  logicOf,
  procOf,
  decorateLogic,
  delay,
  task,
  callTask,
  run,
  bind,
  mount,
  hydrate,
  utils,
  config
};
