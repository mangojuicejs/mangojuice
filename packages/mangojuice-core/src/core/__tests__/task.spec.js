import { child, message, logicOf, procOf, task, utils } from 'mangojuice-core';
import { runWithTracking } from 'mangojuice-test';


describe('task', () => {
  it('should call success handler if provided', async () => {
    class TestLogic {
      create() {
        return task(this.testTask)
          .success(this.successHandler)
      }
      *testTask() {
        yield utils.delay(50);
        return { hello: 'there!' };
      }
      successHandler(...args) {
        return { success: args };
      }
    }

    const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });

    expect(app.model).toMatchSnapshot();
    expect(commands).toMatchSnapshot();
  });

  it('should use identity success command by default', async () => {
    class TestLogic {
      create() {
        return task(this.testTask)
      }
      *testTask() {
        yield utils.delay(50);
        return { hello: 'there!' };
      }
    }

    const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });

    expect(app.model).toMatchSnapshot();
    expect(commands).toMatchSnapshot();
  });

  it('should call fail handler if provided', async () => {
    class TestLogic {
      create() {
        return task(this.testTask)
          .fail(this.failHandler)
      }
      *testTask() {
        yield utils.delay(50);
        throw new Error('ooops');
        return { hello: 'there!' };
      }
      failHandler(...args) {
        return { fail: args };
      }
    }

    const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });

    expect(app.model).toMatchSnapshot();
    expect(commands).toMatchSnapshot();
  });

  it('should emit update event yielded from task', async () => {
    class TestLogic {
      create() {
        return task(this.testTask);
      }
      update = jest.fn();
      *testTask() {
        yield message(() => ({ hey: 'there!' }));
      }
    }

    const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
    const logic = logicOf(app.model);

    expect(logic.update.mock.calls).toMatchSnapshot();
  });

  it('should emit update event yielded from subtask', async () => {
    class TestLogic {
      create() {
        return task(this.testTask);
      }
      update = jest.fn();
      *testTask() {
        yield this.testSubtask;
      }
      *testSubtask() {
        yield message(() => ({ hey: 'there!' }));
      }
    }

    const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
    const logic = logicOf(app.model);

    expect(logic.update.mock.calls).toMatchSnapshot();
  });

  it('should cancel a task', async () => {
    const handler = jest.fn();
    class TestLogic {
      create() {
        return task(this.testTask)
          .success(this.successHandler)
          .fail(this.failHandler)
      }
      cancelTask() {
        return task(this.testTask).cancel();
      }
      *testTask() {
        yield utils.delay(100);
        handler();
        return { hello: 'there!' };
      }
      successHandler(...args) {
        return { success: args };
      }
      failHandler(...args) {
        return { fail: args };
      }
    }

    const res = runWithTracking({ app: { Logic: TestLogic } });
    const { app, commands } = res;

    app.proc.exec(logicOf(app.model).cancelTask);
    await app.proc.finished();

    expect(handler).toHaveBeenCalledTimes(0);
    expect(app.model).toMatchSnapshot();
    expect(commands).toMatchSnapshot();
  });

  it('should cancel a task with subtasks', async () => {
    const handler = jest.fn();
    class TestLogic {
      create() {
        return task(this.testTask)
          .success(this.successHandler)
          .fail(this.failHandler)
      }
      cancelTask() {
        return task(this.testTask).cancel();
      }
      *subSubTask() {
        yield utils.delay(100);
        handler();
      }
      *subTask() {
        yield this.subSubTask();
        handler();
      }
      *testTask() {
        yield this.subTask;
        handler();
        return { hello: 'there!' };
      }
      successHandler(...args) {
        return { success: args };
      }
      failHandler(...args) {
        return { fail: args };
      }
    }

    const res = runWithTracking({ app: { Logic: TestLogic } });
    const { app, commands } = res;

    app.proc.exec(logicOf(app.model).cancelTask);
    await app.proc.finished();

    expect(handler).toHaveBeenCalledTimes(0);
    expect(app.model).toMatchSnapshot();
    expect(commands).toMatchSnapshot();
  });

  it('should accept arguments from a task', async () => {
    const handler = jest.fn();
    class TestLogic {
      create() {
        return task(this.testTask)
          .args(42, 'forty-two', [42], { value: 42 }, function () {});
      }
      *testTask(...args) {
        handler(...args);
      }
    }

    const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });

    expect(handler.mock.calls).toMatchSnapshot();
  });

  it('should support fast path with non-generator task function', async () => {
    class TestLogic {
      create() {
        return task(this.testTask)
          .success(this.successHandler)
      }
      testTask() {
        return { hello: 'there!' };
      }
      successHandler(...args) {
        return { success: args };
      }
    }

    const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });

    expect(app.model).toMatchSnapshot();
    expect(commands).toMatchSnapshot();
  });

  it('should handle error in non-generator task function', async () => {
    class TestLogic {
      create() {
        return task(this.testTask)
          .fail(this.failHandler)
      }
      testTask() {
        throw new Error('ooops');
        return { hello: 'there!' };
      }
      failHandler(...args) {
        return { fail: args };
      }
    }

    const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });

    expect(app.model).toMatchSnapshot();
    expect(commands).toMatchSnapshot();
  });

  it('should cancel prev executing task by default', async () => {
    class TestLogic {
      create() {
        return task(this.testTask)
          .success(this.successHandler)
      }
      *testTask() {
        yield utils.delay(50);
        return { hello: (this.model.hello || '') + 'there!' };
      }
      successHandler(...args) {
        return { success: args };
      }
    }

    const { app, commands } = runWithTracking({ app: { Logic: TestLogic } });
    await utils.delay(40);
    app.proc.exec(logicOf(app.model).create);
    await utils.delay(40);
    app.proc.exec(logicOf(app.model).create);
    await app.proc.finished();

    expect(app.model).toMatchSnapshot();
    expect(commands).toMatchSnapshot();
  });

  it('should NOT cancel prev executing task in multithread mode', async () => {
    class TestLogic {
      create() {
        return task(this.testTask)
          .multithread()
          .success(this.successHandler)
      }
      *testTask() {
        yield utils.delay(50);
        return { hello: (this.model.hello || '') + 'there!' };
      }
      successHandler(result) {
        return result;
      }
    }

    const { app, commands } = runWithTracking({ app: { Logic: TestLogic } });
    await utils.delay(40);
    app.proc.exec(logicOf(app.model).create);
    await utils.delay(40);
    app.proc.exec(logicOf(app.model).create);
    await app.proc.finished();

    expect(app.model).toMatchSnapshot();
    expect(commands).toMatchSnapshot();
  });

  describe('yeilded array', () => {
    it('should aggregate several promises', async () => {
      const handler = jest.fn();
      class TestLogic {
        create() {
          return task(this.testTask);
        }
        *testTask(...args) {
          const a = utils.delay(10).then(() => 'a');
          const b = utils.delay(20).then(() => 'b');
          const c = utils.delay(30).then(() => 'c');

          const res = yield [a, b, c];
          handler(res);
        }
      }
      const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
      expect(handler.mock.calls).toMatchSnapshot();
    })

    it('should noop with no args', async () => {
      const handler = jest.fn();
      class TestLogic {
        create() {
          return task(this.testTask);
        }
        *testTask(...args) {
          const res = yield [];
          handler(res);
        }
      }
      const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
      expect(handler.mock.calls).toMatchSnapshot();
    })

    it('should support an array of generators', async () => {
      const handler = jest.fn();
      class TestLogic {
        create() {
          return task(this.testTask);
        }
        *testTask(...args) {
          const res = yield [function*(){ return 1 }()];
          handler(res);
        }
      }
      const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
      expect(handler.mock.calls).toMatchSnapshot();
    })
  });

  describe('yeilded generator', () => {
    it('should wrap children generators', async () => {
      const handler = jest.fn();
      class TestLogic {
        create() {
          return task(this.testTask);
        }
        *work() {
          yield utils.delay(50);
          return 'yay';
        }
        *testTask(...args) {
          const a = yield this.work;
          const b = yield this.work;
          const c = yield this.work;
          handler(a,b,c);

          const res = yield [this.work, this.work, this.work];
          handler(res);
        }
      }
      const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
      expect(handler.mock.calls).toMatchSnapshot();
    })
  });

  describe('yeilded invalid', () => {
    it('shuold throw an error', async () => {
      class TestLogic {
        create() {
          return task(this.testTask);
        }
        *testTask(...args) {
          yield null;
        }
      }

      const { app, errors } = await runWithTracking({
        app: { Logic: TestLogic },
        expectErrors: true
      });

      expect(errors).toMatchSnapshot();
    });
  });

  describe('yeilded object', () => {
    it('should aggregate several promises', async () => {
      const handler = jest.fn();
      class TestLogic {
        create() {
          return task(this.testTask);
        }
        *testTask(...args) {
          const a = utils.delay(10).then(() => 'a');
          const b = utils.delay(20).then(() => 'b');
          const c = utils.delay(30).then(() => 'c');

          const res = yield {
            a: a,
            b: b,
            c: c
          };
          handler(res);
        }
      }
      const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
      expect(handler.mock.calls).toMatchSnapshot();
    })

    it('should noop with no args', async () => {
      const handler = jest.fn();
      class TestLogic {
        create() {
          return task(this.testTask);
        }
        *testTask(...args) {
          const res = yield {};
          handler(res);
        }
      }
      const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
      expect(handler.mock.calls).toMatchSnapshot();
    })

    it('should ignore non-thunkable properties', async () => {
      const handler = jest.fn();
      function Pet(name) {
        this.name = name;
        this.something = function(){};
      }
      class TestLogic {
        create() {
          return task(this.testTask);
        }
        *testTask(...args) {
          const foo = {
            name: { first: 'tobi' },
            age: 2,
            address: utils.delay(10).then(() => 'aaa'),
            tobi: new Pet('tobi'),
            now: new Date(2017,2,2),
            falsey: false,
            nully: null,
            undefiney: undefined,
          };
          const res = yield foo;
          handler(res);
        }
      }
      const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
      expect(handler.mock.calls).toMatchSnapshot();
    })

    it('should preserve key order', async () => {
      const handler = jest.fn();
      class TestLogic {
        create() {
          return task(this.testTask);
        }
        *testTask(...args) {
          const before = {
            a: utils.delay(10).then(() => 'a'),
            b: utils.delay(20).then(() => 'b'),
            c: utils.delay(30).then(() => 'c'),
          };

          const after = yield before;
          handler(
            Object.keys(before).join(','),
            Object.keys(after).join(',')
          );
        }
      }
      const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
      expect(handler.mock.calls).toMatchSnapshot();
    });
  });

  describe('yeilded promise', () => {
    function getPromise(val, err) {
      return new Promise(function (resolve, reject) {
        if (err) reject(err);
        else resolve(val);
      });
    }

    describe('with one promise yield', () => {
      it('should work', async () => {
        const handler = jest.fn();
        class TestLogic {
          create() {
            return task(this.testTask);
          }
          *testTask(...args) {
            const a = yield getPromise(1);
            handler(a);
          }
        }
        const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
        expect(handler.mock.calls).toMatchSnapshot();
      })
    });

    describe('with several promise yields', function(){
      it('should work', async () => {
        const handler = jest.fn();
        class TestLogic {
          create() {
            return task(this.testTask);
          }
          *testTask(...args) {
            const a = yield getPromise(1);
            const b = yield getPromise(2);
            const c = yield getPromise(3);
            handler(a,b,c);
          }
        }
        const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
        expect(handler.mock.calls).toMatchSnapshot();
      })
    });

    describe('when a promise is rejected', () => {
      it('should throw and resume', async () => {
        const handler = jest.fn();
        class TestLogic {
          create() {
            return task(this.testTask);
          }
          *testTask(...args) {
            let error;
            try {
              yield getPromise(1, new Error('boom'));
            } catch (err) {
              error = err;
            }
            handler(error);
          }
        }

        const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
        expect(handler.mock.calls).toMatchSnapshot();
      });
    });
  });

  describe('yielded recsive object/array', () => {
    it('should aggregate arrays within arrays', async () => {
      const handler = jest.fn();
      class TestLogic {
        create() {
          return task(this.testTask);
        }
        *work() {
          yield utils.delay(50);
          return 'yay';
        }
        *testTask(...args) {
          const a = utils.delay(10).then(() => 'a');
          const b = utils.delay(20).then(() => 'b');
          const c = utils.delay(30).then(() => 'c');

          const res = yield [a, [b, c]];
          handler(res);
        }
      }
      const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
      expect(handler.mock.calls).toMatchSnapshot();
    })

    it('should aggregate objects within objects', async () => {
      const handler = jest.fn();
      class TestLogic {
        create() {
          return task(this.testTask);
        }
        *work() {
          yield utils.delay(50);
          return 'yay';
        }
        *testTask(...args) {
          const a = utils.delay(10).then(() => 'a');
          const b = utils.delay(20).then(() => 'b');
          const c = utils.delay(30).then(() => 'c');

          const res = yield {
            0: a,
            1: {
              0: b,
              1: c
            }
          };
          handler(res);
        }
      }
      const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
      expect(handler.mock.calls).toMatchSnapshot();
    })
  });

  describe('yeilded thunk', () => {
    function get(val, err, error) {
      return function(done){
        if (error) throw error;
        setTimeout(function(){
          done(err, val);
        }, 10);
      }
    }

    describe('with one yield', () => {
      it('should work', async () => {
        const handler = jest.fn();
        class TestLogic {
          create() {
            return task(this.testTask);
          }
          *testTask(...args) {
            const a = yield get(1);
            handler(a);
          }
        }

        const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
        expect(handler.mock.calls).toMatchSnapshot();
      })
    })

    describe('with several yields', () => {
      it('should work', async () => {
        const handler = jest.fn();
        class TestLogic {
          create() {
            return task(this.testTask);
          }
          *testTask(...args) {
            const a = yield get(1);
            const b = yield get(2);
            const c = yield get(3);
            handler(a,b,c);
          }
        }

        const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
        expect(handler.mock.calls).toMatchSnapshot();
      })
    })

    describe('with many arguments', () => {
      it('should return an array', async () => {
        function exec(cmd) {
          return function(done){
            done(null, 'stdout', 'stderr');
          }
        }

        const handler = jest.fn();
        class TestLogic {
          create() {
            return task(this.testTask);
          }
          *testTask(...args) {
            const out = yield exec('something');
            handler(out);
          }
        }

        const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
        expect(handler.mock.calls).toMatchSnapshot();
      })
    })

    describe('when the function throws', () => {
      it('should be caught', async () => {
        const handler = jest.fn();
        class TestLogic {
          create() {
            return task(this.testTask);
          }
          *testTask(...args) {
            try {
              var a = yield get(1, null, new Error('boom'));
            } catch (err) {
              handler(err);
            }
          }
        }

        const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
        expect(handler.mock.calls).toMatchSnapshot();
      })
    })

    describe('when an error is passed then thrown', () => {
      it('should only catch the first error only', async () => {
        class TestLogic {
          create() {
            return task(this.testTask)
              .fail(this.failHandler);
          }
          *testTask(...args) {
            yield function (done){
              done(new Error('first'));
              throw new Error('second');
            }
          }
          failHandler(...args) {
            return { fail: args };
          }
        }

        const { app, commands } = await runWithTracking({ app: { Logic: TestLogic } });
        expect(app.model).toMatchSnapshot();
      })
    });
  });
});
