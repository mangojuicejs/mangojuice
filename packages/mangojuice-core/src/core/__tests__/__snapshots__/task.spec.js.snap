// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`task should NOT cancel prev executing task in multithread mode 1`] = `
Object {
  "hello": "there!there!there!",
}
`;

exports[`task should NOT cancel prev executing task in multithread mode 2`] = `
Array [
  TaskCmd {
    "execEvery": true,
    "executor": [Function],
    "successCmd": [Function],
    "task": [Function],
  },
  TaskCmd {
    "execEvery": true,
    "executor": [Function],
    "successCmd": [Function],
    "task": [Function],
  },
  Object {
    "hello": "there!",
  },
  TaskCmd {
    "execEvery": true,
    "executor": [Function],
    "successCmd": [Function],
    "task": [Function],
  },
  Object {
    "hello": "there!there!",
  },
  Object {
    "hello": "there!there!there!",
  },
]
`;

exports[`task should accept arguments from a task 1`] = `
Array [
  Array [
    42,
    "forty-two",
    Array [
      42,
    ],
    Object {
      "value": 42,
    },
    [Function],
  ],
]
`;

exports[`task should call fail handler if provided 1`] = `
Object {
  "fail": Array [
    [Error: ooops],
  ],
}
`;

exports[`task should call fail handler if provided 2`] = `
Array [
  TaskCmd {
    "executor": [Function],
    "failCmd": [Function],
    "task": [Function],
  },
  Object {
    "fail": Array [
      [Error: ooops],
    ],
  },
]
`;

exports[`task should call success handler if provided 1`] = `
Object {
  "success": Array [
    Object {
      "hello": "there!",
    },
  ],
}
`;

exports[`task should call success handler if provided 2`] = `
Array [
  TaskCmd {
    "executor": [Function],
    "successCmd": [Function],
    "task": [Function],
  },
  Object {
    "success": Array [
      Object {
        "hello": "there!",
      },
    ],
  },
]
`;

exports[`task should cancel a task 1`] = `Object {}`;

exports[`task should cancel a task 2`] = `
Array [
  TaskCmd {
    "executor": [Function],
    "failCmd": [Function],
    "successCmd": [Function],
    "task": [Function],
  },
  TaskCmd {
    "cancelTask": true,
    "executor": [Function],
    "task": [Function],
  },
]
`;

exports[`task should cancel a task with subtasks 1`] = `Object {}`;

exports[`task should cancel a task with subtasks 2`] = `
Array [
  TaskCmd {
    "executor": [Function],
    "failCmd": [Function],
    "successCmd": [Function],
    "task": [Function],
  },
  TaskCmd {
    "cancelTask": true,
    "executor": [Function],
    "task": [Function],
  },
]
`;

exports[`task should cancel prev executing task by default 1`] = `
Object {
  "success": Array [
    Object {
      "hello": "there!",
    },
  ],
}
`;

exports[`task should cancel prev executing task by default 2`] = `
Array [
  TaskCmd {
    "executor": [Function],
    "successCmd": [Function],
    "task": [Function],
  },
  TaskCmd {
    "executor": [Function],
    "successCmd": [Function],
    "task": [Function],
  },
  TaskCmd {
    "executor": [Function],
    "successCmd": [Function],
    "task": [Function],
  },
  Object {
    "success": Array [
      Object {
        "hello": "there!",
      },
    ],
  },
]
`;

exports[`task should emit update event yielded from subtask 1`] = `
Array [
  Array [
    Message {
      "hey": "there!",
    },
  ],
]
`;

exports[`task should emit update event yielded from task 1`] = `
Array [
  Array [
    Message {
      "hey": "there!",
    },
  ],
]
`;

exports[`task should handle error in non-generator task function 1`] = `
Object {
  "fail": Array [
    [Error: ooops],
  ],
}
`;

exports[`task should handle error in non-generator task function 2`] = `
Array [
  TaskCmd {
    "executor": [Function],
    "failCmd": [Function],
    "task": [Function],
  },
  Object {
    "fail": Array [
      [Error: ooops],
    ],
  },
]
`;

exports[`task should support fast path with non-generator task function 1`] = `
Object {
  "success": Array [
    Object {
      "hello": "there!",
    },
  ],
}
`;

exports[`task should support fast path with non-generator task function 2`] = `
Array [
  TaskCmd {
    "executor": [Function],
    "successCmd": [Function],
    "task": [Function],
  },
  Object {
    "success": Array [
      Object {
        "hello": "there!",
      },
    ],
  },
]
`;

exports[`task should use identity success command by default 1`] = `
Object {
  "hello": "there!",
}
`;

exports[`task should use identity success command by default 2`] = `
Array [
  TaskCmd {
    "executor": [Function],
    "task": [Function],
  },
  Object {
    "hello": "there!",
  },
]
`;

exports[`task yeilded array should aggregate several promises 1`] = `
Array [
  Array [
    Array [
      "a",
      "b",
      "c",
    ],
  ],
]
`;

exports[`task yeilded array should noop with no args 1`] = `
Array [
  Array [
    Array [],
  ],
]
`;

exports[`task yeilded array should support an array of generators 1`] = `
Array [
  Array [
    Array [
      1,
    ],
  ],
]
`;

exports[`task yeilded generator should wrap children generators 1`] = `
Array [
  Array [
    "yay",
    "yay",
    "yay",
  ],
  Array [
    Array [
      "yay",
      "yay",
      "yay",
    ],
  ],
]
`;

exports[`task yeilded invalid shuold throw an error 1`] = `
Array [
  [TypeError: You may only yield a function, promise, generator, array, object or message but the following object was passed: "null"],
]
`;

exports[`task yeilded object should aggregate several promises 1`] = `
Array [
  Array [
    Object {
      "a": "a",
      "b": "b",
      "c": "c",
    },
  ],
]
`;

exports[`task yeilded object should ignore non-thunkable properties 1`] = `
Array [
  Array [
    Object {
      "address": "aaa",
      "age": 2,
      "falsey": false,
      "name": Object {
        "first": "tobi",
      },
      "now": 2017-03-01T19:00:00.000Z,
      "nully": null,
      "tobi": Pet {
        "name": "tobi",
        "something": [Function],
      },
      "undefiney": undefined,
    },
  ],
]
`;

exports[`task yeilded object should noop with no args 1`] = `
Array [
  Array [
    Object {},
  ],
]
`;

exports[`task yeilded object should preserve key order 1`] = `
Array [
  Array [
    "a,b,c",
    "a,b,c",
  ],
]
`;

exports[`task yeilded promise when a promise is rejected should throw and resume 1`] = `
Array [
  Array [
    [Error: boom],
  ],
]
`;

exports[`task yeilded promise with one promise yield should work 1`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`task yeilded promise with several promise yields should work 1`] = `
Array [
  Array [
    1,
    2,
    3,
  ],
]
`;

exports[`task yeilded thunk when an error is passed then thrown should only catch the first error only 1`] = `
Object {
  "fail": Array [
    [Error: first],
  ],
}
`;

exports[`task yeilded thunk when the function throws should be caught 1`] = `
Array [
  Array [
    [Error: boom],
  ],
]
`;

exports[`task yeilded thunk with many arguments should return an array 1`] = `
Array [
  Array [
    "stdout",
  ],
]
`;

exports[`task yeilded thunk with one yield should work 1`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`task yeilded thunk with several yields should work 1`] = `
Array [
  Array [
    1,
    2,
    3,
  ],
]
`;

exports[`task yielded recsive object/array should aggregate arrays within arrays 1`] = `
Array [
  Array [
    Array [
      "a",
      Array [
        "b",
        "c",
      ],
    ],
  ],
]
`;

exports[`task yielded recsive object/array should aggregate objects within objects 1`] = `
Array [
  Array [
    Object {
      "0": "a",
      "1": Object {
        "0": "b",
        "1": "c",
      },
    },
  ],
]
`;
